### DISCUSSION TITLE
How to enable version tracking in the configuration of Apisix to achieve the effect of config as code?

### DISCUSSION THREAD

**Discussion by flylan** [apache/apisix#10855] [Q&A] [answered]
As far as I know, the current configuration of Apisix either chooses to store all in yaml or all in etcd, which cannot achieve mixed storage.

So how do most companies that currently apply Apisix to production environments solve the problem of deploying Apisix from grayscale testing to production?

Is it true that most companies simply manually copy configurations or simulate curls, republishing grayscale configurations to the production environment?

I feel like I haven't found a reliable way to deploy configuration from a grayscale environment to a production environment yet. According to the previous way of using nginx, it is basically to directly modify the nginx configuration in the production environment, with few grayscale tests. Basically, if there is a manual error in operation and maintenance, it will directly cause the production environment to crash

Are we really resorting to primitive methods to manage the transition of gateway configurations from the staging to the production environment, now that we are using apisix?

One solution that I have come up with temporarily is to save the configuration on Git and update Apisix through the CICD tool to reduce human errors during the process. Is there any other better solution?

**Comment by juzhiyuan** [comment, marked as answer]
> One solution that I have come up with temporarily is to save the configuration on Git and update Apisix through the CICD tool to reduce human errors during the process. Is there any other better solution?

Yes. CI/CD is a good way to publish configurations across different environments. And I suggest using [ADC CLI](https://docs.api7.ai/apisix/reference/adc) to do that. One example with ADC and APISIX-based Software: https://docs.api7.ai/enterprise/best-practices/devops-adc#sync-configuration

Also, you can consider using the latest Ingress Controller in CI/CD, as `v2.0+` removes the dependency of ETCD. Refer to https://github.com/apache/apisix/issues/11338#issuecomment-3162166205
